### PR TITLE
Show upstream/downstreams from a dataset

### PR DESCRIPTION
**Pull request by bbovenzi**
Datasets list: show a count of upstreams and downstreams for a dataset
Dataset details: show a full list of upstreams and downstreams for the dataset, and each item links to the DAG's grid view

Found a bug in how we auto-generated API types where arrays weren't transformed correctly. Fixed this by importing a better type transformation function instead of us having a custom one.

We'll refine the UI in a subsequent PR but this at least displays the data.

Screen Shot 2022-08-01 at 10 05 53 AM
Screen Shot 2022-08-01 at 10 05 47 AM

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/www/static/js/datasets/Details.tsx [hunk 1]
```diff
@@ -85,32 +147,11 @@ const DatasetDetails = ({ datasetId, onBack }: Props) => {
...
-            <Text mr={1}>Created At:</Text>
-            <Time dateTime={dataset.createdAt} />
-          </Flex>
-        </Box>
-      )}
-      <Heading size="lg" mt={3} mb={2} fontWeight="normal">Upstream Events</Heading>
-      <Text>Whenever a DAG has updated this dataset.</Text>
+      {!!dataset && (<Details dataset={dataset} />)}
+      <Flex alignItems="center">
+        <Heading size="lg" mt={3} mb={2} fontWeight="normal">Upstream Events</Heading>
```
Comment (dstandish) [inline airflow/www/static/js/datasets/Details.tsx]:
```suggestion
        <Heading size="lg" mt={3} mb={2} fontWeight="normal">Events</Heading>
```

they are just events, in this context.

File: airflow/www/static/js/datasets/Details.tsx [hunk 2]
```diff
@@ -30,12 +30,74 @@ import {
...
+      </Flex>
+    )}
+    <Flex my={2}>
+      <Text mr={1}>Updated At:</Text>
+      <Time dateTime={updatedAt} />
+    </Flex>
+    <Flex my={2}>
+      <Text mr={1}>Created At:</Text>
+      <Time dateTime={createdAt} />
+    </Flex>
```
Comment (dstandish) [inline airflow/www/static/js/datasets/Details.tsx]:
```suggestion
```

In this context, adding updated and created at are maybe a bit confusing.  Cus "updated at" could indicated "the dataset was _updated_" i.e. in an event.

File: airflow/www/static/js/datasets/Details.tsx [hunk 3]
```diff
@@ -30,12 +30,74 @@ import {
...
+        <Link key={`${dagId}.${taskId}`} color="blue.600" href={`/dags/${dagId}/grid`}>
+          {`${dagId}.${taskId}`}
+        </Link>
+      ))}
+    </Box>
+    )}
+    {downstreamDagReferences && !!downstreamDagReferences.length && (
+    <Box>
+      <Flex alignItems="center">
+        <Heading size="md" fontWeight="normal">Downstream DAGs</Heading>
```
Comment (dstandish) [inline airflow/www/static/js/datasets/Details.tsx]: i think you might need some `<br />` tags somewhere ... they are all on same line
image

File: airflow/www/static/js/datasets/Details.tsx (hunk 2 above)
Comment (bbovenzi) [reply airflow/www/static/js/datasets/Details.tsx]: Oh that isn't the case, its just when the db entry was updated? Ok. I'll cut that out.

**Comment by ashb** [comment]
Thought/idea: Rather than "upstream" and "downstream" how about we use "producing" and "consuming"?

**Comment by potiuk** [comment]
> Thought/idea: Rather than "upstream" and "downstream" how about we use "producing" and "consuming"?

I very much like this proposal. This is so much more "real-live" than downstream/upstream. Maybe it's the non-native English  kicking-in, but upstream/downstream has no 'natural" mental model for me when it comes to data flows (counterintuitvely you could say).

I have to always think twice what it really means, Producer/Consumer is very straightforward.

**Review by ashb — APPROVED** [review]

**Pull request merged**